src/lib/hookify/vanillaHooks.ts

```diff
@@ -0,0 +1,349 @@
+      if (!fx.has(hook)) fx.set(hook, new Set());
+      state.push({
+        type: 'externalStore',
+        value: getSnapshot(),
+        getSnapshot,
+        subRef: undefined as any, // overrideing immediately below
+        doSub: undefined as any, // handled below
+        cleanup: undefined,
+        cleaned: false,
+        hook,
+      });
+    }
+
+    const thisState = state[thisHookInfo.index] as SyncedExternalStoreEffect;
+    thisState.getSnapshot = getSnapshot;
+    if (subscribe !== thisState.subRef) {
+      thisState.subRef = subscribe;
+      thisState.doSub = () => {
+        let killed = false;
+        const doUnsub = subscribe(() => {
+          if (killed) return;
+          const newVal = thisState.getSnapshot();
+          if (newVal !== thisState.value) {
+            thisState.value = newVal;
+            reschedule(thisHookInfo);
+          }
+        });
+        return () => {
+          killed = true;
+          doUnsub();
+        };
+      };
+      // queue sub for async execution
+      effects.push(thisState);
+    }
+
+    // eslint-disable-next-line no-plusplus
+    return (state[thisHookInfo.index++] as SyncedExternalStoreEffect).value;
+  };
+
+  const useReducer = <S, A>(
+    reducer: (state: S, action: A) => S,
+    initialState: S | (() => S),
+    init?: (arg: S | (() => S)) => S
+  ): [S, (action: A) => void] => {
+    const thisHookInfo = getInfo()!;
+    const { index, state } = thisHookInfo;
+    if (index === state.length)
+      state.push({
+        type: 'state',
+        value: isFunction(init) ? init(initialState) : getValue(undefined, initialState),
+        set: (value: A) => {
+          const ef = state[index] as StateEffect;
+          const newVal = reducer(ef.value as S, value);
+          if (ef.value !== newVal) {
+            ef.value = newVal;
+            reschedule(thisHookInfo);
+          }
+        },
+      });
+    // eslint-disable-next-line no-plusplus
+    const { value, set } = state[thisHookInfo.index++] as StateEffect;
+    return [value, set];
+  };
+
+  const useState = <S>(
+    initialState: S | (() => S)
+  ): [S, (newState: S | ((prevState: S) => S)) => void] =>
+    useReducer((_, value) => getValue(_, value), initialState);
+
+  const useRef = <T>(initialValue: T): { current: T } => {
+    const thisHookInfo = getInfo()!;
+    const { index, state } = thisHookInfo;
+    if (index === state.length) {
+      state.push({ type: 'ref', val: { current: initialValue } });
+    }
+    // eslint-disable-next-line no-plusplus
+    return (state[thisHookInfo.index++] as RefEffect).val;
+  };
+
+  return {
+    createContext,
+    dropEffect,
+    hasEffect,
+    hooked,
+    basicRerunHook,
+    useCallback,
+    useContext,
+    useEffect,
+    useLayoutEffect,
+    useMemo,
+    useReducer,
+    useRef,
+    useState,
+    useSyncExternalStore,
+    wait,
+  };
+})();
+
+export default hookifyHooks;
+export const useCallbackHf = hookifyHooks.useCallback;
```

Comment:
what does the `Hf` in all of these exports stand for?

Reply:
Yep! Can rename this is just sort of what I ended up doing for clarity/shortness.